s3torchbenchmarking/src/s3torchbenchmarking/datagen.py

```diff
@@ -0,0 +1,380 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  // SPDX-License-Identifier: BSD
+import abc
+import io
+import os
+import tarfile
+import time
+from dataclasses import dataclass
+from multiprocessing import Queue
+from pathlib import Path
+from threading import Thread, Lock, Barrier
+from typing import Any, Callable, Iterator, TypeVar, Tuple, Dict, List, Optional, Union
+
+import boto3
+import click
+import numpy as np
+import prefixed as pr
+import yaml
+from PIL import Image
+
+
+@dataclass
+class LabelledSample:
+    label: str
+    data: io.BytesIO
+
+
+class Sentinel:
+    pass
+
+
+class DataGenerator(abc.ABC):
+    @abc.abstractmethod
+    def create(self, idx_gen: Iterator[int]) -> Iterator[LabelledSample]:
+        pass
+
+
+class ImageGenerator(DataGenerator):
+    def __init__(self, width: int, height: int, image_format: str):
+        self.width = width
+        self.height = height
+        self.img_format = image_format
+
+    def create(self, idx_iter: Iterator[int]) -> Iterator[LabelledSample]:
+        for index in idx_iter:
+            img = Image.fromarray(
+                np.random.randint(
+                    0, high=256, size=(self.height, self.width, 3), dtype=np.uint8
+                )
+            )
+            byte_buf = io.BytesIO()
+            img.save(byte_buf, format=self.img_format)
+            byte_buf.seek(0)
+
+            yield LabelledSample(
+                label=f"image_{index}.{self.img_format}", data=byte_buf
+            )
+
+
+class ThreadSafeCounter:
+    def __init__(self) -> None:
+        self.val = 0
+        self.lock = Lock()
+
+    def get_and_inc(self):
+        with self.lock:
+            self.val += 1
+            return self.val
+
+
+class Utils:
```

Comment:
Not blocking: Why is this in a class? Most (all?) of these functions would work perfectly well at the top level